### ISSUE TITLE
Segmentation Fault (Core Dumped) Bug in Fine-Tuned YOLOv8 Model with Low Confidence Inference

### ISSUE THREAD

**Issue by takhyun12** [ultralytics/ultralytics#11881]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Predict

### Bug

I would like to share a significant bug related to confidence inferences identified in the fine-tuned YOLOv8 model.

I built a custom dataset through Roboflow and fine-tuned it using YOLOv8x.

I discovered that the fine-tuned model only triggers a Segmentation Fault (core dumped) when inferring specific images in a Linux environment. I analyzed this issue using various profiling tools.

As a result, I found a close relationship between the conf parameter and this issue, as detailed below:

First, I am sharing the fine-tuned model and test image I used in my experiments:

model : [download](https://drive.google.com/file/d/1a1bxCgTg5CaU3p-rO1ODcGR9Sjkbelqc/view?usp=sharing)

image : [download](https://drive.google.com/file/d/1Vfa4cGeaBMH9Xld85uKmqKsb6b9ZRjMr/view?usp=sharing)

My findings are as follows:

- The issue does not occur when the conf parameter is not used or set to the default value of 0.25.
- For the attached model, a conf value lower than 0.15 results in a Segmentation Fault for specific images.
- Lower conf values cause Segmentation Faults with different images, but for the same parameter value, the issue is reproducible with the same images.

I tried various experiments to resolve this issue, but none were effective:

- Testing after converting the model to TorchScript, ONNX, and OpenVINO
- Using torch.no_grad(), torch.inference_mode(), gc.collect(), and torch.empty_cache()
- Using NGC Docker and Ultralytics Docker
- Training in different GPU environments such as A100, H100, RTX4090, and multi-GPU setups
- Testing various parameters like stream, half, and cpu operations during inference

*This issue is consistently reproducible in all Linux environments*, but it does not occur in Windows. I believe that when the conf - value is low, there might be an explosive increase in memory usage during the model inference.

I wrote test code to infer around 1,000 images and have been running it. If the problem does not reproduce with the images I have attached, it could be due to subtle changes in dependencies such as torch.

Please run about 100 images through my fine-tuned model with a low conf value, and you'll quickly find the problematic images.

I created a model that performs best with a low conf value of 0.01 or 0.001, but I'm unable to use it due to this issue.

I would greatly appreciate your assistance.

### Environment

```
Ultralytics YOLOv8.1.9 🚀 Python-3.10.13 torch-2.0.1+cu117 CUDA:0 (NVIDIA A100 80GB PCIe, 81229MiB)
Setup complete ✅ (128 CPUs, 251.5 GB RAM, 2729.3/3080.1 GB disk)

OS                  Linux-4.18.0-425.19.2.el8_7.x86_64-x86_64-with-glibc2.31
Environment         Docker
Python              3.10.13
Install             pip
RAM                 251.48 GB
CPU                 AMD EPYC 7543 32-Core Processor
CUDA                11.7

matplotlib          ✅ 3.8.3>=3.3.0
numpy               ✅ 1.26.4>=1.22.2
opencv-python       ✅ 4.8.0.76>=4.6.0
pillow              ✅ 10.2.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.12.0>=1.4.1
torch               ✅ 2.0.1>=1.8.0
torchvision         ✅ 0.15.2>=0.9.0
tqdm                ✅ 4.66.1>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.2.0>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
```

### Minimal Reproducible Example

```python
from ultralytics import YOLO

# Use the model I've attached.
model = YOLO('fine_tuned_model.pt')

# Use the image I've attached and set the conf parameter to a low value as follows.
model.predict('image.jpg', conf=0.01)
````

### Additional

I think the following PRs are closely related to the issue I'm reporting:

- https://github.com/ultralytics/ultralytics/issues/2473
- https://github.com/ultralytics/ultralytics/issues/9898

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
Hello! Thank you for the detailed report and for conducting such thorough investigations into this issue. 🙌

The Segmentation Fault you're experiencing when setting a low confidence threshold is indeed intriguing. It sounds like an issue related to memory handling or buffer overflow that's specific to the configurations in a Linux environment, as you suggested.

Given that you've already tried multiple environments and configurations without success, here is a potential workaround:
- **Limit the output detection**: Sometimes setting a very low confidence threshold can yield an overwhelmingly large number of detections per image, which might be causing memory overload. You could initially try with a higher `conf`, and incrementally lower it to find a safe threshold that avoids the Segmentation Fault.

Additionally, considering this seems related to how Linux handles memory versus Windows, examining the system calls and memory management differences in your Docker or Linux environment could provide more insights. Using tools like `valgrind` might help identify if there's specific illegal memory access occurring.

Could you please check if limiting the number of detections resolves the issue? Meanwhile, I'll further investigate potential fixes from our side as well.
  
Thank you for offering to help with a PR! We appreciate your contribution. Let's aim to isolate this behavior further and work on a solid patch. 🚀

**Comment by takhyun12** [comment]
@glenn-jocher 

I conducted experiments to limit the number of detections as you suggested.
I set the `conf` to 0.01 and varied `max_det` as demonstrated in the code below:

```python
with torch.inference_mode():
    segment_results = self.model.predict(
        source=image,
        save=False,
        save_txt=False,
        imgsz=(640, 640),
        max_det={max_detection},
        conf=0.01,
        stream=True,
    )
```

#### Experiment Results
- Based on the attached model, it was confirmed that no errors occurred when reducing `max_det` to 2 or less.
- Errors occurred independently of the inference time.

| `max_det` | Result                   | Console Output                                                 | Time  |
|-----------|--------------------------|--------------------------------------------------------|-------|
| 1         | ok                       | -                                                      | -     |
| 2         | ok                       | -                                                      | -     |
| 3         | Segmentation fault       | 640x640: 1 crow's feet, 1 mental crease, 1 nasolabial fold | 90.0ms |
| 5         | Segmentation fault       | 640x640: 4 crow's feet, 1 mental crease                | 84.0ms |
| 10        | Segmentation fault       | 640x640: 7 crow's feet, 1 mental crease, 2 nasolabial folds | 37.8ms |
| 100       | Segmentation fault       | 640x640: 53 crow's feet, 14 forehead wrinkles, 1 frown line, 13 marionette lines, 7 mental creases, 12 nasolabial folds | 27.1ms |


#### Environment
- There are no other processes running on my server besides this one.

```
Ultralytics YOLOv8.2.11 🚀 Python-3.10.14 torch-2.3.0+cu121 CUDA:0 (NVIDIA A100-SXM4-80GB, 81053MiB)
Setup complete ✅ (128 CPUs, 251.8 GB RAM, 612.1/12469.0 GB disk)

OS                  Linux-4.19.93-1.nbp.el7.x86_64-x86_64-with-glibc2.35
Environment         Docker
Python              3.10.14
Install             pip
RAM                 251.82 GB
CPU                 AMD EPYC 7543 32-Core Processor
CUDA                12.1

matplotlib          ✅ 3.8.4>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.13.0>=1.4.1
torch               ✅ 2.3.0>=1.8.0
torchvision         ✅ 0.18.0>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
```

Please let me know if there is anything else you need.

**Comment by takhyun12** [comment]
I've discovered an interesting point that if I comment out the conf, no errors occur no matter how high max_det is set.

```python
with torch.inference_mode():
    segment_results = self.model.predict(
        source=image,
        save=False,
        save_txt=False,
        imgsz=(640, 640),
        max_det=100,
        # conf=0.01,
        stream=True,
    )
```
My trained model performs exceptionally well when the conf is low. Is there another method to achieve the same effect without using the conf option?

**Comment by glenn-jocher** [comment]
@takhyun12 hello! Thanks for the interesting find! 😊 It's quite insightful that omitting the `conf` parameter entirely avoids the segmentation fault issue.

Regarding achieving effective low confidence operation without explicitly setting a very low `conf` value, one approach is to filter the results post-prediction based on your confidence threshold. This way, you can use a default or slightly lower `conf` setting that does not cause crashes, and then programmatically discard detections below your desired confidence level in the post-processing step:

```python
results = model.predict(source=image, imgsz=(640, 640), max_det=100, stream=True)
filtered_results = [r for r in results if r.confidence >= 0.01]
```

This method allows you to handle an arbitrary number of detections more safely and gives you the flexibility to adjust your confidence threshold dynamically after observing model output.

Let me know if this works for you or if further adjustments are needed!

**Comment by takhyun12** [comment]
@glenn-jocher Hello,

I have an additional question regarding this issue.

The code you suggested did not solve the problem:

```python
results = model.predict(source=image, imgsz=(640, 640), max_det=100, stream=True)
filtered_results = [r for r in results if r.confidence >= 0.01]
```

The results of the above code are completely different from the following code:

```python
results = model.predict(source=image, imgsz=(640, 640), conf=0.01, max_det=100, stream=True)
```

Output:

| Output Image A | Output Image B (use `conf`) |
|:--------------:|:--------------:|
| ![a](https://github.com/ultralytics/ultralytics/assets/41291493/b15ef854-73b4-46c4-a86a-d7947d5dc721) | ![b](https://github.com/ultralytics/ultralytics/assets/41291493/3eef79bc-38c4-4968-a35b-de66fdda908c) |


I think there might be a difference in how confidence is handled during prediction and post-processing. Do you have any suggestions for this case? Or is there any progress in fixing this Segmentation Fault bug?

Thank you for your help.

**Comment by glenn-jocher** [comment]
Hello @takhyun12,

Thank you for your follow-up and for testing out the code snippet. You're correct in observing that filtering post-prediction can yield different results compared to setting the `conf` threshold directly in the `predict` method. This difference arises because setting `conf` directly affects the model's internal scoring and non-max suppression process during prediction, which can't be exactly replicated by filtering afterwards.

As for the segmentation fault issue, it's still under investigation. A potential workaround could be to adjust the `conf` parameter slightly higher to a stable threshold that doesn't trigger the fault, and then fine-tune the `max_det` parameter to manage the number of detections.

Here's a slight modification to try:

```python
results = model.predict(source=image, imgsz=(640, 640), conf=0.05, max_det=100, stream=True)
filtered_results = [r for r in results if r.confidence >= 0.01]
```

This approach uses a slightly higher internal confidence while still allowing you to capture lower confidence detections for further analysis.

Let's stay in touch as we work through these issues. Your feedback is invaluable! 🌟

**Comment by takhyun12** [comment]
Hello @glenn-jocher ,

I wanted to share some findings from my research and seek your assistance regarding an issue I've encountered. 
I've identified that a segmentation fault is occurring within the `cv2.abi3.so` library when running the YOLOv8.

Here are the details from the `gdb` backtrace, showing that the segmentation fault occurs in `cv2.abi3.so`:

```
0: 640x480 2 crows_feets, 4 nasolabial_foldss, 4 neck_wrinkless, 20.9ms Speed: 0.9ms preprocess, 20.9ms inference, 2.0ms postprocess per image at shape (1, 3, 640, 480) 
0: 640x512 3 crows_feets, 1 mental_crease, 5 nasolabial_foldss, 1 neck_wrinkles, 26.0ms Speed: 0.9ms preprocess, 26.0ms inference, 1.5ms postprocess per image at shape (1, 3, 640, 512) 
0: 640x480 5 crows_feets, 1 mental_crease, 3 nasolabial_foldss, 1 neck_wrinkles, 22.9ms Speed: 0.8ms preprocess, 22.9ms inference, 3.6ms postprocess per image at shape (1, 3, 640, 480) 
Thread 1 "python" received signal SIGSEGV, Segmentation fault. 0x00007f42de945e71 in ?? () from /opt/conda/envs/firm/lib/python3.10/site-packages/cv2/cv2.abi3.so
```

Using `valgrind`, I determined that the segmentation fault results from `cv2.abi3.so` attempting to access the invalid memory address `0xfffffffffffffff0`.

Here is the relevant `valgrind` output:
```
0: 640x480 32 crows_feets, 1 frown_line, 15 marionette_lines, 11 nasolabial_foldss, 46 neck_wrinkless, 382.7ms
Speed: 29.6ms preprocess, 382.7ms inference, 57.8ms postprocess per image at shape (1, 3, 640, 480)
==12192== Invalid read of size 8
==12192==    at 0x5AB59391: ??? (in /opt/conda/envs/firm/lib/python3.10/site-packages/cv2/cv2.abi3.so)
==12192==    by 0x5AB5E0BA: ??? (in /opt/conda/envs/firm/lib/python3.10/site-packages/cv2/cv2.abi3.so)
==12192==    by 0x5AB5E4F2: ??? (in /opt/conda/envs/firm/lib/python3.10/site-packages/cv2/cv2.abi3.so)
==12192==    by 0x5A70D736: ??? (in /opt/conda/envs/firm/lib/python3.10/site-packages/cv2/cv2.abi3.so)
==12192==    by 0x4FDC86: cfunction_call (methodobject.c:543)
==12192==    by 0x4F741A: _PyObject_MakeTpCall (call.c:215)
==12192==    by 0x4F34C5: UnknownInlinedFun (abstract.h:112)
==12192==    by 0x4F34C5: UnknownInlinedFun (abstract.h:99)
==12192==    by 0x4F34C5: UnknownInlinedFun (abstract.h:123)
==12192==    by 0x4F34C5: UnknownInlinedFun (ceval.c:5893)
==12192==    by 0x4F34C5: _PyEval_EvalFrameDefault (ceval.c:4181)
==12192==    by 0x4FE0CE: UnknownInlinedFun (pycore_ceval.h:46)
==12192==    by 0x4FE0CE: UnknownInlinedFun (ceval.c:5067)
==12192==    by 0x4FE0CE: _PyFunction_Vectorcall (call.c:342)
==12192==    by 0x4EF4A2: UnknownInlinedFun (abstract.h:114)
==12192==    by 0x4EF4A2: UnknownInlinedFun (abstract.h:123)
==12192==    by 0x4EF4A2: UnknownInlinedFun (ceval.c:5893)
==12192==    by 0x4EF4A2: _PyEval_EvalFrameDefault (ceval.c:4231)
==12192==    by 0x5099CD: UnknownInlinedFun (pycore_ceval.h:46)
==12192==    by 0x5099CD: UnknownInlinedFun (ceval.c:5067)
==12192==    by 0x5099CD: UnknownInlinedFun (call.c:342)
==12192==    by 0x5099CD: UnknownInlinedFun (abstract.h:114)
==12192==    by 0x5099CD: method_vectorcall (classobject.c:53)
==12192==    by 0x4EF4A2: UnknownInlinedFun (abstract.h:114)
==12192==    by 0x4EF4A2: UnknownInlinedFun (abstract.h:123)
==12192==    by 0x4EF4A2: UnknownInlinedFun (ceval.c:5893)
==12192==    by 0x4EF4A2: _PyEval_EvalFrameDefault (ceval.c:4231)
==12192==    by 0x4FE0CE: UnknownInlinedFun (pycore_ceval.h:46)
==12192==    by 0x4FE0CE: UnknownInlinedFun (ceval.c:5067)
==12192==    by 0x4FE0CE: _PyFunction_Vectorcall (call.c:342)
==12192==  Address 0xfffffffffffffff0 is not stack'd, malloc'd or (recently) free'd
==12192== 
==12192== 
==12192== Process terminating with default action of signal 11 (SIGSEGV)
==12192==  Access not within mapped region at address 0xFFFFFFFFFFFFFFF0
==12192==    at 0x5AB59391: ??? (in /opt/conda/envs/firm/lib/python3.10/site-packages/cv2/cv2.abi3.so)
==12192==    by 0x5AB5E0BA: ??? (in /opt/conda/envs/firm/lib/python3.10/site-packages/cv2/cv2.abi3.so)
==12192==    by 0x5AB5E4F2: ??? (in /opt/conda/envs/firm/lib/python3.10/site-packages/cv2/cv2.abi3.so)
==12192==    by 0x5A70D736: ??? (in /opt/conda/envs/firm/lib/python3.10/site-packages/cv2/cv2.abi3.so)
==12192==    by 0x4FDC86: cfunction_call (methodobject.c:543)
==12192==    by 0x4F741A: _PyObject_MakeTpCall (call.c:215)
==12192==    by 0x4F34C5: UnknownInlinedFun (abstract.h:112)
==12192==    by 0x4F34C5: UnknownInlinedFun (abstract.h:99)
==12192==    by 0x4F34C5: UnknownInlinedFun (abstract.h:123)
==12192==    by 0x4F34C5: UnknownInlinedFun (ceval.c:5893)
==12192==    by 0x4F34C5: _PyEval_EvalFrameDefault (ceval.c:4181)
==12192==    by 0x4FE0CE: UnknownInlinedFun (pycore_ceval.h:46)
==12192==    by 0x4FE0CE: UnknownInlinedFun (ceval.c:5067)
==12192==    by 0x4FE0CE: _PyFunction_Vectorcall (call.c:342)
==12192==    by 0x4EF4A2: UnknownInlinedFun (abstract.h:114)
==12192==    by 0x4EF4A2: UnknownInlinedFun (abstract.h:123)
==12192==    by 0x4EF4A2: UnknownInlinedFun (ceval.c:5893)
==12192==    by 0x4EF4A2: _PyEval_EvalFrameDefault (ceval.c:4231)
==12192==    by 0x5099CD: UnknownInlinedFun (pycore_ceval.h:46)
==12192==    by 0x5099CD: UnknownInlinedFun (ceval.c:5067)
==12192==    by 0x5099CD: UnknownInlinedFun (call.c:342)
==12192==    by 0x5099CD: UnknownInlinedFun (abstract.h:114)
==12192==    by 0x5099CD: method_vectorcall (classobject.c:53)
==12192==    by 0x4EF4A2: UnknownInlinedFun (abstract.h:114)
==12192==    by 0x4EF4A2: UnknownInlinedFun (abstract.h:123)
==12192==    by 0x4EF4A2: UnknownInlinedFun (ceval.c:5893)
==12192==    by 0x4EF4A2: _PyEval_EvalFrameDefault (ceval.c:4231)
==12192==    by 0x4FE0CE: UnknownInlinedFun (pycore_ceval.h:46)
==12192==    by 0x4FE0CE: UnknownInlinedFun (ceval.c:5067)
==12192==    by 0x4FE0CE: _PyFunction_Vectorcall (call.c:342)
==12192==  If you believe this happened as a result of a stack
==12192==  overflow in your program's main thread (unlikely but
==12192==  possible), you can try to increase the size of the
==12192==  main thread stack using the --main-stacksize= flag.
==12192==  The main thread stack size used in this run was 8388608.
==12192== 
==12192== HEAP SUMMARY:
==12192==     in use at exit: 3,784,020,939 bytes in 1,699,091 blocks
==12192==   total heap usage: 4,354,347 allocs, 2,655,256 frees, 9,667,277,435 bytes allocated
```

The invalid memory address `0xfffffffffffffff0` corresponds to `rsi`:

```
(gdb) info registers
rax            0x0                 0
rbx            0x0                 0
rcx            0x10                16
rdx            0x1                 1
rsi            0xfffffffffffffff0  -16
rdi            0x0                 0
rbp            0x0                 0x0
rsp            0x7ffe8acb1220      0x7ffe8acb1220
r8             0x7ffe8acb1370      140731226985328
r9             0x8                 8
r10            0x0                 0
r11            0x7ffe8acb1350      140731226985296
r12            0x0                 0
r13            0x0                 0
r14            0x7ffe8acb1520      140731226985760
r15            0x7ffe8acb1390      140731226985360
rip            0x7f4874a4a391      0x7f4874a4a391
eflags         0x10256             [ PF AF ZF IF RF ]
cs             0x33                51
ss             0x2b                43
ds             0x0                 0
es             0x0                 0
fs             0x0                 0
gs             0x0                 0
```

The line causing the error is as follows:

```
(gdb) disassemble /r 0x7f4874a4a391-50, 0x7f4874a4a391+50
Dump of assembler code from 0x7f4874a4a35f to 0x7f4874a4a3c3:
   0x00007f4874a4a35f:  00 00   add    %al,(%rax)
   0x00007f4874a4a361:  48 8b bc 24 d8 00 00 00 mov    0xd8(%rsp),%rdi
   0x00007f4874a4a369:  2b 94 24 d0 00 00 00    sub    0xd0(%rsp),%edx
   0x00007f4874a4a370:  48 89 4c 24 30  mov    %rcx,0x30(%rsp)
   0x00007f4874a4a375:  89 d1   mov    %edx,%ecx
   0x00007f4874a4a377:  89 54 24 14     mov    %edx,0x14(%rsp)
   0x00007f4874a4a37b:  ba 01 00 00 00  mov    $0x1,%edx
   0x00007f4874a4a380:  48 63 df        movslq %edi,%rbx
   0x00007f4874a4a383:  48 c1 ff 20     sar    $0x20,%rdi
   0x00007f4874a4a387:  4c 89 44 24 20  mov    %r8,0x20(%rsp)
   0x00007f4874a4a38c:  48 89 5c 24 08  mov    %rbx,0x8(%rsp)
=> 0x00007f4874a4a391:  48 03 1e        add    (%rsi),%rbx
   0x00007f4874a4a394:  49 89 dc        mov    %rbx,%r12
   0x00007f4874a4a397:  44 89 4c 24 10  mov    %r9d,0x10(%rsp)
   0x00007f4874a4a39c:  49 d3 e4        shl    %cl,%r12
   0x00007f4874a4a39f:  0f b6 8c 24 d0 00 00 00 movzbl 0xd0(%rsp),%ecx
   0x00007f4874a4a3a7:  d3 e2   shl    %cl,%edx
   0x00007f4874a4a3a9:  d1 fa   sar    %edx
   0x00007f4874a4a3ab:  01 fa   add    %edi,%edx
   0x00007f4874a4a3ad:  4c 89 df        mov    %r11,%rdi
   0x00007f4874a4a3b0:  48 63 da        movslq %edx,%rbx
   0x00007f4874a4a3b3:  48 89 5c 24 18  mov    %rbx,0x18(%rsp)
   0x00007f4874a4a3b8:  48 03 5e 08     add    0x8(%rsi),%rbx
   0x00007f4874a4a3bc:  49 8b 73 08     mov    0x8(%r11),%rsi
   0x00007f4874a4a3c0:  49 2b 33        sub    (%r11),%rsi
End of assembler dump.
```

In summary, it appears that `cv2.abi3.so`, which is part of the opencv-python package installed with `pip install ultralytics`, is attempting to access an invalid memory address (0xfffffffffffffff0). This issue doesn't occur on Windows OS, likely due to different memory handling mechanisms.

I've tried the following solutions without success:
- Reinstalling opencv-python
- Manually building and installing opencv-python on the Linux environment
- Any assistance or insights on how to resolve this issue would be greatly appreciated.

Thank you for your help.
